vkrunner/vr-box.c

```diff
@@ -331,36 +323,67 @@ compare_cb(enum vr_box_base_type base_type,
                            base_type,
                            (const uint8_t *) data->a + offset,
                            (const uint8_t *) data->b + offset,
-                           data->tolerance,
-                           data->tolerance_in_percent)) {
+                           data->tolerance)) {
                 data->result = false;
                 return false;
         }
 
         return true;
 }
 
+static void
+convert_float_to_double(double *dst, const float *src, int n) {
+        for (int i = 0; i < n; ++i)
+                dst[i] = (double) src[i];
+}
+
 bool
 vr_box_compare(enum vr_box_comparison comparison,
-               double tolerance,
-               bool percent,
+               const struct vr_script_tolerance *tolerance,
                const struct vr_box *a,
                const struct vr_box *b)
 {
         assert(a->type == b->type);
 
-        struct compare_closure data = {
-                .comparison = comparison,
-                .a = a,
-                .b = b,
-                .tolerance = tolerance,
-                .tolerance_in_percent = percent,
-                .result = true
-        };
-
-        vr_box_for_each_component(a->type, compare_cb, &data);
-
-        return data.result;
+        const struct vr_box_type_info *info = type_infos + a->type;
+        if ((info->base_type == VR_BOX_BASE_TYPE_FLOAT ||
+             info->base_type == VR_BOX_BASE_TYPE_DOUBLE) &&
+            comparison == VR_BOX_COMPARISON_FUZZY_EQUAL) {
```

Comment:
It would be nice if this could reuse the `vr_box_for_each_component` function instead of duplicating it here. Perhaps the compare closure could just have an “index” member which just counts the number of times `compare_cb` is called. Then if `vr_tolerance_equal` took a `component` argument instead of `num_components` like I originally suggested then `compare_cb` could just pass `data->index % num_rows` for the component number. That would also mean we wouldn’t have to have this `convert_float_to_double` helper.

Reply:
I couldn't think of it. Thank you for the suggestion.